### PR TITLE
BUG: fix pointer arithmetic in _get_field_view

### PR DESCRIPTION
**Pull request by ahaldane**
Should have used PyArray_BYTES, not PyArray_DATA

1.10 fix coming soon, hopefully...

### CODE REVIEW

**Comment by ahaldane** [comment]
I'll fix the warnings too, in a minute

**Comment by ahaldane** [comment]
All right, hopefully that fixes warnings. At least, I don't see warnings in the build log, but I'm not sure I saw them before either. 

Also I'm not 100% sure about the last cast to int, maybe @seberg can confirm that's right.

File: numpy/core/src/multiarray/mapping.c [hunk 1]
```diff
@@ -169,7 +169,7 @@ prepare_index(PyArrayObject *self, PyObject *index,
     int new_ndim, fancy_ndim, used_ndim, index_ndim;
     int curr_idx, get_idx;
 
-    npy_intp i, n;
+    int i, n;
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: This doesn't look right as `PySequence_Size` is later assigned to `n`
```
Py_ssize_t PySequence_Size(PyObject *o)
```
`npy_intp` is effectively the same as `Py_ssize_t`. Note that one needs to be *very* careful in replaceing `npy_intp` by `int`.

**Comment by seberg** [comment]
I think I did some dual use at some points maybe, or maybe it is not, and there is simply a < MAX_DIMS check? Some of the dual use could maybe also just be avoided. The cast to int is correct, I have no clue why it is not an int in the first place. A permutation is an axes, and axes are numbered by ints.

File: numpy/core/src/multiarray/mapping.c (hunk 1 above)
Comment (seberg) [inline numpy/core/src/multiarray/mapping.c]: Yeah, n needs to be intp, but i is probably fine. Though would need to check more carefully that it is only used to iterate something like axes/dimensions.

**Comment by charris** [comment]
Add 1.10.2 release tag for reference.

**Comment by astrofrog** [comment]
This fixes the original issue from #6701, but I now run into more errors (see https://github.com/numpy/numpy/issues/6701#issuecomment-158427400) - not sure if they should be addressed in this PR though?

**Comment by ahaldane** [comment]
Yes that looks like something we can have a separate PR for. 

For this PR, I've looked updated it more carefully now. I'm still not 100% sure we want to be silencing these warnings, because maybe the correct fix is to change the type of `value` in `npy_index_info`, and the type of `perm` in `npy_stride_sort_item`. But looking at the code they both get assigned to both ints and intps, so it's not clear.

**Comment by seberg** [comment]
I think npy_stride_sort_item is public API probably, so cannot change. The `value` field is used as intp sometimes, but we could split that up into two fields (which would probably be cleaner)....

**Comment by astrofrog** [comment]
@ahaldane @seberg - side note, @charris pointed me to https://github.com/numpy/numpy/pull/6662/files which apparently has a slightly different fix for the mapping.c file

**Comment by seberg** [comment]
Yeah, I like the BYTES more, but I think we can probably just leave it away here and merge the other in any case?

**Comment by charris** [comment]
I like the bytes more also ;) I'm not going to backport #6662 as it contains fixes for things not in 1.10, so the fix against 1.10.x is still needed.

**Comment by charris** [comment]
I went ahead and merged #6662, so this needs a rebase.

**Comment by ahaldane** [comment]
Rebased and took a stab at splitting up the 'value' field. It avoids any casting but adds a bit more complexity, let me know if you prefer the fix from 2nd commit, or if you see a better way.

File: numpy/core/src/multiarray/mapping.h [hunk 2]
```diff
@@ -16,11 +16,17 @@ typedef struct {
...
-     * boolean array if it is a converted boolean array.
+     * For HAS_INTEGER, int_index is the value of the integer index. For
+     * HAS_ELLIPSIS, ellipsis_slices is the number of slices the ellipsis is
+     * worth. For HAS_FANCY, fancy_bool is -1 for an integer array, or the
+     * length of the array for boolean arrays.
      */
-    npy_intp value;
+    union{
+        npy_intp int_index;
+        int ellipsis_slices;
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/mapping.h]: We are never using the union to convert between types, so behavior does not depend on machine endianess, right?

File: numpy/core/src/multiarray/mapping.h [hunk 3]
```diff
@@ -16,11 +16,17 @@ typedef struct {
...
-     * Value of an integer index, number of slices an Ellipsis is worth,
-     * -1 if input was an integer array and the original size of the
-     * boolean array if it is a converted boolean array.
+     * For HAS_INTEGER, int_index is the value of the integer index. For
+     * HAS_ELLIPSIS, ellipsis_slices is the number of slices the ellipsis is
+     * worth. For HAS_FANCY, fancy_bool is -1 for an integer array, or the
+     * length of the array for boolean arrays.
      */
-    npy_intp value;
+    union{
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/mapping.h]: [This](http://stackoverflow.com/questions/3228104/anonymous-union-within-struct-not-in-c99) suggest that anonymous unions will only work with GCC or very recent compilers. Can anyone check if this works with MSVC 2008?

**Comment by ahaldane** [comment]
Thanks for reviewing @jaimefrio. I got rid of the union, you're right it's only a C11 feature. (It was a premature optimization anyway)

**Comment by jaimefrio** [comment]
You could always give the union parameter a name, e.g. `value`, and access the fields as e,g, `.value.int_index`. Your lines would be longer, but the code easier to follow. Finding a syntax that works on all compilers we support aside, I think using a union was a very good idea.

**Comment by ahaldane** [comment]
Done. That was a good idea, I think it's even clearer this way than with an anonymous union.

File: numpy/core/src/multiarray/mapping.h [hunk 4]
```diff
@@ -16,11 +16,17 @@ typedef struct {
...
+     * HAS_ELLIPSIS, ellipsis_slices is the number of slices the ellipsis is
+     * worth. For HAS_FANCY, fancy_bool is -1 for an integer array, or the
+     * length of the array for boolean arrays.
      */
-    npy_intp value;
+    union {
+        npy_intp int_index;
+        int ellipsis_slices;
+        npy_intp fancy_bool;
+    } value;
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.h]: `value` is too generic, a more descriptive name would help. That said, this just looks strange. I assume that the aim is to have a single value type for everything, but this is bound to be confusing, just having three different types might be preferable, i.e., `npy_int_index`, `npy_ellipsis_slice`, or something like that. The use of unions is something of a red flag.

**Comment by charris** [comment]
This is getting pretty complex. For 1.10.2, what would be the minimal fix. @seberg Would it be safe to ignore the warnings and just fix the pointer?

**Comment by njsmith** [comment]
The existing overloading of the value field does seem a bit overly clever, but it should work even without all this union stuff, because a npy_intp should always be at least as large as an int...

**Comment by ahaldane** [comment]
All right, how about we just go with the first two commits here and drop the rest? The only thing the remaining commits saves us from are two casts from intp to int in `mapiter_fill_info`, and those are safe casts anyway.

I removed the union-related commits. If you still want to see them they are [here](https://github.com/ahaldane/numpy/commits/fix_getfield_p_arith_splitvalue)

**Comment by charris** [comment]
@njsmith, the union is likely to cause problems on big endian machines also. It's why I like little endian.

File: numpy/core/src/multiarray/mapping.c [hunk 5]
```diff
@@ -348,14 +349,14 @@ prepare_index(PyArrayObject *self, PyObject *index,
 #else
             if (PyLong_CheckExact(obj) || !PyArray_Check(obj)) {
 #endif
-                i = PyArray_PyIntAsIntp(obj);
-                if ((i == -1) && PyErr_Occurred()) {
+                npy_intp ind = PyArray_PyIntAsIntp(obj);
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: Blank line after.

File: numpy/core/src/multiarray/mapping.c [hunk 6]
```diff
@@ -527,15 +528,15 @@ prepare_index(PyArrayObject *self, PyObject *index,
                  * sure that array-likes or odder arrays are always
                  * handled right.
                  */
-                i = PyArray_PyIntAsIntp((PyObject *)arr);
+                npy_intp ind = PyArray_PyIntAsIntp((PyObject *)arr);
```
Comment (charris) [inline numpy/core/src/multiarray/mapping.c]: Blank line after.

**Comment by charris** [comment]
LGTM, just a couple of nits.

**Comment by seberg** [comment]
Yeah, the union was better then before but I guess there is no good reason to be clever here, I agree I was silly to pack everything into the same value in the first place. 
Of course the warnings were safe to ignore.

**Comment by ahaldane** [comment]
Updated with last suggestions

**Comment by charris** [comment]
Thanks Allan.

**Comment by ahaldane** [comment]
Thanks everyone for reviews & suggestions.

**Pull request merged**